src/els_db.erl

```diff
@@ -1,133 +1,93 @@
 -module(els_db).
 
 %% API
--export([ find/2
-        , find_multi/2
-        , keys/1
-        , list/1
-        , store/3
-        , update/4
+-export([ clear_table/1
+        , clear_tables/0
         , delete/2
-        , flush_all_tables/0
-        , start_link/0
-        ]).
-
-%% gen_server callbacks
--export([ init/1
-        , handle_call/3
-        , handle_cast/2
+        , install/0
+        , install/1
+        , lookup/2
+        , match/1
+        , transaction/1
+        , wait_for_tables/0
+        , wait_for_tables/1
+        , write/1
         ]).
 
 -define(SERVER, ?MODULE).
--define(TABLES, [ {documents,  set}
-                , {modules,    set}
-                , {references, bag}
-                , {signatures, set}
+-define(TABLES, [ els_dt_module
+                , els_dt_documents
+                , els_dt_references
+                , els_dt_signatures
                 ]).
-
--type state() :: #{}.
--type table() :: atom().
--type key()   :: any().
+-define(TIMEOUT, 5000).
 
 %%==============================================================================
 %% Exported functions
 %%==============================================================================
 
--spec start_link() -> {ok, pid()}.
-start_link() ->
-  gen_server:start_link({local, ?SERVER}, ?MODULE, {}, []).
-
--spec find(table(), key()) -> {ok, any()} | {error, not_found}.
-find(Table, Key) ->
-  case ets:lookup(Table, Key) of
-    [] -> {error, not_found};
-    [{Key, Value}] -> {ok, Value}
-  end.
-
--spec find_multi(table(), key()) -> {ok, nonempty_list(any())}
-                                    | {error, not_found}.
-find_multi(Table, Key) ->
-  case ets:lookup(Table, Key) of
-    [] -> {error, not_found};
-    KVs -> {ok, KVs} %% Return as is, we need to iterate the result
-                     %% anyway, no point doing it twice.
-  end.
-
--spec keys(table()) -> [any()].
-keys(Table) ->
-  [K || {K, _} <- ets:tab2list(Table)].
-
--spec list(table()) -> [any()].
-list(Table) ->
-  ets:tab2list(Table).
-
--spec store(table(), key(), any()) -> ok.
-store(Table, Key, Value) ->
-  ets:insert(Table, {Key, Value}),
+-spec install() -> ok.
+install() ->
+  {ok, Dir} = application:get_env(erlang_ls, db_dir),
+  install(Dir).
+
+-spec install(string()) -> ok.
+install(Dir) ->
+  lager:info("Creating DB. [dir=~s]", [Dir]),
+  ok = filelib:ensure_dir(filename:join([Dir, "dummy"])),
+  ok = application:set_env(mnesia, dir, Dir),
+  mnesia:create_schema([node()]),
+  application:start(mnesia),
+  create_tables(),
+  application:stop(mnesia).
+
+-spec create_tables() -> ok.
+create_tables() ->
+  [els_db_table:create(T) || T <- ?TABLES],
   ok.
 
--spec update(table(), key(), function(), any()) -> ok.
-update(Table, Key, UpdateFun, Default) ->
```

Comment:
It turns out this function was unused?

Reply:
We stopped using it when we changes the `references` table to a `bag`.